### ISSUE TITLE
[TS2.5 regression] TypeScript complains about functions returning never on some paths

### ISSUE THREAD

**Issue by mprobst** [Microsoft/TypeScript#21102]
**TypeScript Version:**  2.6.2

**Code**

```ts
function neverReturn(): never { throw 1; }
function f() {
  if (Math.random() > 0.5) {
    return neverReturn();
  }
}
```

**Expected behavior:**

In 2.5, this compiled fine. The return type of `f` is inferred as `void`.

**Actual behavior:**

In 2.6, this complains:

```
Not all code paths return a value.
```

The return type of `f` is inferred as `undefined`.

**Comment by mhegazy** [comment]
This seems correct to me. a behavior change nevertheless.
